File: markov.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef int bool;
#define TRUE 1
#define FALSE 0

size_t M = 7;

void flip(int *lst, int left, int right) {
  while (left < right) {
	int tmp = -lst[left];
	lst[left] = -lst[right - 1];
	lst[right - 1] = tmp;
	left++;
	right--;
  }
}

void rand_flip(int *lst) {
  int left, right;
  left = rand() / (RAND_MAX / (M + 1) + 1);
  right = rand() % (M + 1);

  if (left > right) {
	int tmp = left;
	left = right;
	right = left;
  }

  flip(lst, left, right);
}

void shuffle_and_flip(int *lst) {
  for (size_t i = 0; i < M; i++) {
	size_t j = i + rand() / (RAND_MAX / (M - i) + 1);
	int tmp = lst[i];
	lst[i] = lst[j];
	lst[j] = tmp;

	if (rand() % 2 == 1)
	  lst[i] = -lst[i];
  }
}


int main() {
  size_t particle_count = 10000000;
  int *particles, *start;

  srand(time(NULL));

  particles = malloc(particle_count * M * sizeof(int));
  start = malloc(M * sizeof(int));

  for (size_t j = 0; j < M; j++) {
	start[j] = j + 1;
  }
  shuffle_and_flip(start);

  for (size_t i = 0; i < particle_count; i++) {
	memcpy(particles + i * M, start, M * sizeof(int));
	/* for (size_t j = 0; j < M; j++) { */
	/*   particles[i * M + j] = j + 1; */
	/* } */
	/* shuffle_and_flip(particles + i * M); */
  }

  free(start);

  for (size_t iter = 1; iter < 100; iter++) {
	size_t count = 0;

	printf("Iteration %u ... ", iter);

	for (size_t i = 0; i < particle_count; i++) {
	  bool correct = TRUE;

	  rand_flip(particles + i * M);

	  for (size_t j = 0; j < M; j++) {
		if (particles[i * M + j] != j + 1) {
		  correct = FALSE;
		  break;
		}
	  }

	  if (correct)
		count++;
	}

	printf("Ratio: %f %%\n", ((float) count) / particle_count * 100);
  }

  free(particles);

  return 0;
}

File: markov.py
#!/usr/bin/env python

from __future__ import division

import sys
import numpy as np
import random

M = None

def _rand_flip(lst):
    hix_1 = random.randint(0, M)
    hix_2 = random.randint(0, M)

    if hix_1 > hix_2:
        hix_1, hix_2 = hix_2, hix_1

    return _flip(lst, hix_1, hix_2)

def _flip(lst, left, right):
    output = []

    for index in xrange(M):
        if index >= left and index < right:
            output.append(-lst[right - (index - left + 1)])
        else:
            output.append(lst[index])

    return output

def _randomize(lst):
    random.shuffle(lst)
    for index, val in enumerate(lst):
        lst[index] = random.choice([-1, 1]) * val
    return lst

def _main(args):
    if len(args) < 1:
        print "Usage: python {} edge_count".format(__file__)
        sys.exit(1)

    global M
    M = int(args[0])
    particle_count = 10 ** 7

    start = range(1, M + 1)
    correct = range(1, M + 1)
    random.shuffle(start)

    for index, val in enumerate(start):
        start[index] = val * random.choice([-1, 1])

    # print start
    # particles = np.array([_randomize(range(M)) for i in xrange(particle_count)])
    particles = np.array([start for i in xrange(particle_count)])

    iterations = 100
    for i in xrange(iterations):
        print "Iteration {:3} ...".format(i),

        count = 0

        for index in xrange(particle_count):
            particles[index] = _rand_flip(particles[index])
            if all(i == j for i, j in zip(particles[index], correct)):
                count += 1

        print "Correct:", count / particle_count

if __name__ == "__main__":
    _main(sys.argv[1:])
